sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala

```diff
@@ -30,7 +33,7 @@ import org.apache.spark.sql.catalyst.expressions.Expression
  *
  * Implementations should throw [[NoSuchDatabaseException]] when databases don't exist.
  */
-abstract class ExternalCatalog {
+abstract class ExternalCatalog(conf: SparkConf, hadoopConf: Configuration) {
```

Comment:
@cloud-fan  I found it that if we add a parameter `defaultDB` for `ExternalCatalog` and its subclass `InMemoryCatalog` and `HiveExternalCatalog`, this change will cause a lot of related code to be modified, such as test cases ,and other logic which create `InMemoryCatalog` and `HiveExternalCatalog`, for example:  currently all the parameters of `InMemoryCatalog` have its own default value `class InMemoryCatalog(conf: SparkConf = new SparkConf,hadoopConfig: Configuration = new Configuration)`,we can create it without an parameters, but if we add a `defaultDB`, we should new a defaultDB in the parameter, while we can not create a legal deafultDB because we can not get the warehouse path for the defaultDB like this:
` `class InMemoryCatalog(conf: SparkConf = new SparkConf,hadoopConfig: Configuration = new Configuration, defaultDB: CatalogDatabase = CatalogDatabase("default","","${can not get the warehouse path}",Map.empty))``

if we don't provide a default value for defautDB in the parameter, this will cause more code change which I think it is not proper.

what about we keep the `provided def warehousePath` in `ExternalCatalog`, and add a 
`lazy val defaultDB = {
    val qualifiedWarehousePath = SessionCatalog
      .makeQualifiedPath(warehousePath, hadoopConf).toString
   CatalogDatabase("default","", qualifiedWarehousePath, Map.empty)
}` 

this can also avoid call getDatabase